File: src/domain/balanceCalculator.ts
import type { Debt, Group, UserId } from "./models";
import { addCents, assertValidAmountCents } from "./money";

export type BalanceMap = Record<UserId, number>;

export type BalanceCalculatorErrorCode =
  | "invalidPayer"
  | "invalidParticipant"
  | "negativeSplit"
  | "paidByMismatch"
  | "splitMismatch"
  | "nonZeroResidual";

export class BalanceCalculatorError extends Error {
  constructor(
    public readonly code: BalanceCalculatorErrorCode,
    public readonly expenseTitle?: string,
  ) {
    super(createBalanceErrorMessage(code, expenseTitle));
    this.name = "BalanceCalculatorError";
  }
}

export function calculateBalances(group: Group): BalanceMap {
  validateGroupForBalances(group);

  const balances: BalanceMap = Object.fromEntries(group.users.map((user) => [user.id, 0]));

  for (const expense of group.expenses) {
    for (const payer of expense.paidBySplits) {
      balances[payer.participantId] = addCents(balances[payer.participantId] ?? 0, payer.amountCents);
    }

    for (const split of expense.splits) {
      balances[split.participantId] = addCents(balances[split.participantId] ?? 0, -split.amountCents);
    }
  }

  const residual = addCents(...Object.values(balances));
  if (residual !== 0) {
    throw new BalanceCalculatorError("nonZeroResidual");
  }

  return balances;
}

export function calculateDebts(group: Group): Debt[] {
  const balances = calculateBalances(group);

  const creditors = Object.entries(balances)
    .filter(([, amountCents]) => amountCents > 0)
    .map(([id, amountCents]) => ({ id, amountCents }))
    .sort(compareCreditors);

  const debtors = Object.entries(balances)
    .filter(([, amountCents]) => amountCents < 0)
    .map(([id, amountCents]) => ({ id, amountCents }))
    .sort(compareDebtors);

  const debts: Debt[] = [];
  let debtorIndex = 0;
  let creditorIndex = 0;

  while (debtorIndex < debtors.length && creditorIndex < creditors.length) {
    const debtor = debtors[debtorIndex];
    const creditor = creditors[creditorIndex];
    const transferCents = Math.min(-debtor.amountCents, creditor.amountCents);

    if (transferCents > 0) {
      debts.push({
        id: `${debtor.id}-${creditor.id}-${transferCents}`,
        from: debtor.id,
        to: creditor.id,
        amountCents: transferCents,
      });
    }

    debtor.amountCents += transferCents;
    creditor.amountCents -= transferCents;

    if (debtor.amountCents === 0) {
      debtorIndex += 1;
    }
    if (creditor.amountCents === 0) {
      creditorIndex += 1;
    }
  }

  return debts;
}

export function validateGroupForBalances(group: Group): void {
  const userIds = new Set(group.users.map((user) => user.id));

  for (const expense of group.expenses) {
    assertValidAmountCents(expense.amountCents);

    let paidTotal = 0;
    for (const payer of expense.paidBySplits) {
      if (!userIds.has(payer.participantId)) {
        throw new BalanceCalculatorError("invalidPayer", expense.title);
      }
      if (payer.amountCents < 0) {
        throw new BalanceCalculatorError("negativeSplit", expense.title);
      }
      paidTotal = addCents(paidTotal, payer.amountCents);
    }

    let splitTotal = 0;
    for (const split of expense.splits) {
      if (!userIds.has(split.participantId)) {
        throw new BalanceCalculatorError("invalidParticipant", expense.title);
      }
      if (split.amountCents < 0) {
        throw new BalanceCalculatorError("negativeSplit", expense.title);
      }
      splitTotal = addCents(splitTotal, split.amountCents);
    }

    if (paidTotal !== expense.amountCents) {
      throw new BalanceCalculatorError("paidByMismatch", expense.title);
    }
    if (splitTotal !== expense.amountCents) {
      throw new BalanceCalculatorError("splitMismatch", expense.title);
    }
  }
}

function compareCreditors(
  left: { id: string; amountCents: number },
  right: { id: string; amountCents: number },
): number {
  if (left.amountCents === right.amountCents) {
    return left.id.localeCompare(right.id);
  }
  return right.amountCents - left.amountCents;
}

function compareDebtors(
  left: { id: string; amountCents: number },
  right: { id: string; amountCents: number },
): number {
  if (left.amountCents === right.amountCents) {
    return left.id.localeCompare(right.id);
  }
  return left.amountCents - right.amountCents;
}

function createBalanceErrorMessage(code: BalanceCalculatorErrorCode, expenseTitle?: string): string {
  const suffix = expenseTitle ? ` for ${expenseTitle}` : "";

  switch (code) {
    case "invalidPayer":
      return `Expense payer is not part of the group${suffix}.`;
    case "invalidParticipant":
      return `Expense participant is not part of the group${suffix}.`;
    case "negativeSplit":
      return `Expense split cannot be negative${suffix}.`;
    case "paidByMismatch":
      return `Paid-by total does not match expense amount${suffix}.`;
    case "splitMismatch":
      return `Expense total and split total do not match${suffix}.`;
    case "nonZeroResidual":
      return "Balances do not sum to zero.";
  }
}

File: BalanceCalculator.swift
import Foundation

enum BalanceCalculatorError: Error, LocalizedError {
    case invalidPayer
    case invalidParticipant
    case negativeSplit
    case paidByMismatch(expenseTitle: String)
    case splitMismatch(expenseTitle: String)
    case nonZeroResidual

    var errorDescription: String? {
        switch self {
        case .invalidPayer:
            return "Expense payer is not part of the group."
        case .invalidParticipant:
            return "Expense participant is not part of the group."
        case .negativeSplit:
            return "Expense split cannot be negative."
        case .paidByMismatch(let expenseTitle):
            return "Paid-by total does not match expense amount for \(expenseTitle)."
        case .splitMismatch(let expenseTitle):
            return "Expense total and split total do not match for \(expenseTitle)."
        case .nonZeroResidual:
            return "Balances do not sum to zero."
        }
    }
}

enum Money {
    static let scale = 2
    static let zero = Decimal(0)

    static func rounded(_ value: Decimal) -> Decimal {
        var source = value
        var result = Decimal()
        NSDecimalRound(&result, &source, scale, .bankers)
        return result
    }

    static func abs(_ value: Decimal) -> Decimal {
        value < zero ? -value : value
    }

    static func isZero(_ value: Decimal) -> Bool {
        abs(value) <= Decimal(string: "0.005", locale: Locale(identifier: "en_US_POSIX")) ?? Decimal(0.005)
    }
}

enum BalanceCalculator {
    static func calculateBalances(for group: Group) throws -> [UUID: Decimal] {
        try validate(group)

        var balances: [UUID: Decimal] = Dictionary(uniqueKeysWithValues: group.users.map { ($0.id, Money.zero) })

        for expense in group.expenses {
            for payer in expense.paidBySplits {
                balances[payer.userId, default: Money.zero] = Money.rounded(
                    balances[payer.userId, default: Money.zero] + payer.amount
                )
            }

            for split in expense.splits {
                balances[split.userId, default: Money.zero] = Money.rounded(
                    balances[split.userId, default: Money.zero] - split.amount
                )
            }
        }

        let total = Money.rounded(balances.values.reduce(Money.zero, +))
        guard Money.isZero(total) else {
            throw BalanceCalculatorError.nonZeroResidual
        }

        return balances
    }

    static func calculateDebts(for group: Group) throws -> [Debt] {
        let balances = try calculateBalances(for: group)
        // Sorting keeps output deterministic for the same input set.
        var creditors = balances
            .filter { $0.value > Money.zero }
            .map { (id: $0.key, amount: Money.rounded($0.value)) }
            .sorted { lhs, rhs in
                if lhs.amount == rhs.amount {
                    return lhs.id.uuidString < rhs.id.uuidString
                }
                return lhs.amount > rhs.amount
            }

        var debtors = balances
            .filter { $0.value < Money.zero }
            .map { (id: $0.key, amount: Money.rounded($0.value)) }
            .sorted { lhs, rhs in
                if lhs.amount == rhs.amount {
                    return lhs.id.uuidString < rhs.id.uuidString
                }
                return lhs.amount < rhs.amount
            }

        var result: [Debt] = []
        var i = 0
        var j = 0

        while i < debtors.count && j < creditors.count {
            let debtor = debtors[i]
            let creditor = creditors[j]

            let transfer = Money.rounded(min(-debtor.amount, creditor.amount))
            if transfer > Money.zero {
                result.append(Debt(from: debtor.id, to: creditor.id, amount: transfer))
            }

            debtors[i].amount = Money.rounded(debtors[i].amount + transfer)
            creditors[j].amount = Money.rounded(creditors[j].amount - transfer)

            if Money.isZero(debtors[i].amount) {
                i += 1
            }
            if Money.isZero(creditors[j].amount) {
                j += 1
            }
        }

        return result
    }

    static func validate(_ group: Group) throws {
        let userIDs = Set(group.users.map(\.id))

        for expense in group.expenses {
            var paidTotal = Money.zero
            for payer in expense.paidBySplits {
                guard userIDs.contains(payer.userId) else {
                    throw BalanceCalculatorError.invalidPayer
                }
                guard payer.amount >= Money.zero else {
                    throw BalanceCalculatorError.negativeSplit
                }
                paidTotal = Money.rounded(paidTotal + payer.amount)
            }

            var splitTotal = Money.zero
            for split in expense.splits {
                guard userIDs.contains(split.userId) else {
                    throw BalanceCalculatorError.invalidParticipant
                }
                guard split.amount >= Money.zero else {
                    throw BalanceCalculatorError.negativeSplit
                }
                splitTotal = Money.rounded(splitTotal + split.amount)
            }

            let roundedAmount = Money.rounded(expense.amount)
            guard paidTotal == roundedAmount else {
                throw BalanceCalculatorError.paidByMismatch(expenseTitle: expense.title)
            }
            guard splitTotal == roundedAmount else {
                throw BalanceCalculatorError.splitMismatch(expenseTitle: expense.title)
            }
        }
    }
}
